packager/bindlepublisher.go

```diff
@@ -0,0 +1,225 @@
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/deislabs/go-bindle/client"
+	"github.com/deislabs/go-bindle/keyring"
+	"github.com/deislabs/go-bindle/types"
+	"github.com/pelletier/go-toml"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	"github.com/suborbital/subo/project"
+	"github.com/suborbital/subo/subo/util"
+)
+
+const (
+	bindlePublishJobType = "bindle"
+	suboAuthor           = "Subo <subo@suborbital.dev>"
+)
+
+type BindlePublishJob struct{}
+
+type parcelWrapper struct {
+	parcel types.Parcel
+	data   []byte
+}
+
+// NewBindlePublishJob returns a new PublishJob for Bindle.
+func NewBindlePublishJob() PublishJob {
+	b := &BindlePublishJob{}
+
+	return b
+}
+
+// Type returns the publish job's type.
+func (b *BindlePublishJob) Type() string {
+	return bindlePublishJobType
+}
+
+// Publish publishes the application.
+func (b *BindlePublishJob) Publish(log util.FriendlyLogger, ctx *project.Context) error {
+	if ctx.Directive == nil {
+		return errors.New("🚫 cannot push without Directive.yaml file")
+	}
+
+	log.LogStart(fmt.Sprintf("pushing %s@%s", ctx.Directive.Identifier, ctx.Directive.AppVersion))
+
+	invoice := &types.Invoice{
+		BindleVersion: "1.0.0",
+		Bindle: types.BindleSpec{
+			Name:    ctx.Directive.Identifier,
+			Version: strings.TrimPrefix(ctx.Directive.AppVersion, "v"),
+			Authors: []string{
+				suboAuthor,
+			},
+		},
+		Parcel: []types.Parcel{},
+	}
+
+	parcelsBySHA := map[string]parcelWrapper{}
+
+	// add the Directive as a parcel.
+	directiveBytes, err := yaml.Marshal(ctx.Directive)
+	if err != nil {
+		return errors.Wrap(err, "failed to Marshal Directive")
+	}
+
+	directiveParcel := parcelForData("Directive.yaml", "application/yaml", directiveBytes)
+
+	invoice.Parcel = append(invoice.Parcel, directiveParcel)
+
+	parcelsBySHA[directiveParcel.Label.SHA256] = parcelWrapper{
+		parcel: directiveParcel,
+		data:   directiveBytes,
+	}
+
+	// add each Runnable as a parcel.
+	for _, r := range ctx.Runnables {
+		//delete target or .build folder.
+		files, err := ioutil.ReadDir(r.Fullpath)
+		if err != nil {
+			return errors.Wrapf(err, "failed to ReadDir for %s", r.Name)
+		}
+
+		for _, file := range files {
+			if !strings.HasSuffix(file.Name(), ".wasm") {
+				continue
+			}
+
+			fullPath := filepath.Join(r.Fullpath, file.Name())
+
+			fileBytes, err := os.ReadFile(fullPath)
+			if err != nil {
+				return errors.Wrapf(err, "failed to Open %s", fullPath)
+			}
+
+			parcel := parcelForData(file.Name(), "application/wasm", fileBytes)
+
+			invoice.Parcel = append(invoice.Parcel, parcel)
+
+			parcelsBySHA[parcel.Label.SHA256] = parcelWrapper{
+				parcel: parcel,
+				data:   fileBytes,
+			}
+		}
+	}
+
+	sigKey, privKey, err := createOrReadKeypair(suboAuthor)
+	if err != nil {
+		return errors.Wrap(err, "failed to createOrReadKeypair")
+	}
+
+	if err := invoice.GenerateSignature(suboAuthor, types.RoleCreator, sigKey, privKey); err != nil {
```

Comment:
I actually disagree with this case, because in an `if` statement like these, the `err` variable is scoped to that statement and so is actually a new allocation rather than shadowing (which would happen if this was not a one-liner)

For this PR I'm not going to make these changes.